cpp/src/quantiles/tdigest/tdigest.cu

```diff
@@ -0,0 +1,230 @@
+
+  // size of the digest we're querying
+  auto const tdigest_size = tdigest_offsets[tdigest_index + 1] - tdigest_offsets[tdigest_index];
+  double const* cumulative_weight = cumulative_weight_ + tdigest_offsets[tdigest_index];
+
+  // means and weights for the tdigest
+  double const* mean   = mean_ + tdigest_offsets[tdigest_index];
+  double const* weight = weight_ + tdigest_offsets[tdigest_index];
+
+  // if we're at either end
+  double const total_weight = cumulative_weight[tdigest_size - 1];
+  double const cluster_q    = percentage * total_weight;
+  if (cluster_q <= 1) {
+    output[tid] = mean[0];
+    return;
+  } else if (cluster_q >= total_weight - 1) {
+    output[tid] = mean[tdigest_size - 1];
+    return;
+  }
+
+  // otherwise find the centroid we're in and interpolate
+  size_type const centroid_index = static_cast<size_type>(
+    thrust::lower_bound(
+      thrust::seq, cumulative_weight, cumulative_weight + tdigest_size, cluster_q) -
+    cumulative_weight);
+
+  double diff = cluster_q + weight[centroid_index] / 2 - cumulative_weight[centroid_index];
+  if (weight[centroid_index] == 1 && std::abs(diff) < 0.5) {
+    output[tid] = mean[centroid_index];
+    return;
+  }
+  size_type left_index  = centroid_index;
+  size_type right_index = centroid_index;
+  if (diff > 0) {
+    if (right_index == tdigest_size - 1) {
+      output[tid] =
+        lerp(mean[right_index], mean[tdigest_size - 1], diff / (weight[right_index] / 2));
+      return;
+    }
+    right_index++;
+  } else {
+    if (left_index == 0) {
+      output[tid] = lerp(mean[0], mean[left_index], diff / (weight[left_index] / 2));
+      return;
+    }
+    left_index--;
+    diff += weight[left_index] / 2 + weight[right_index] / 2;
+  }
+
+  diff /= (weight[left_index] / 2 + weight[right_index] / 2);
+  output[tid] = lerp(mean[left_index], mean[right_index], diff);
+}
+
+/**
+ * @brief Calculate approximate percentiles on a provided tdigest column.
+ *
+ * Produces a LIST column where each row N represents output from querying the
+ * corresponding tdigest of from row N in `input`. The length of each output list
+ * is the number of percentages specified in `percentages`
+ *
+ * @param input           tdigest input data. One tdigest per row.
+ * @param percentages     Desired percentiles in range [0, 1].
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr              Device memory resource used to allocate the returned column's device
+ * memory
+ *
+ * @returns LIST Column containing requested percentile values.
+ */
+std::unique_ptr<column> compute_approx_percentiles(column_view const& input,
+                                                   column_view const& percentages,
+                                                   rmm::cuda_stream_view stream,
+                                                   rmm::mr::device_memory_resource* mr)
+{
+  lists_column_view lcv(input);
```

Comment:
If the input is intended to be a list column, might as well advertise it in the function signature. 
```suggestion
std::unique_ptr<column> compute_approx_percentiles(lists_column_view const& input,
                                                   column_view const& percentages,
                                                   rmm::cuda_stream_view stream,
                                                   rmm::mr::device_memory_resource* mr)
{
  lists_column_view lcv(input);
```

Reply:
It has since changed to be a struct (had to wrap it in an outer struct to store per-digest min/max values (which are distinct from the min and max centroids within)).  I'll make it a `structs_column_view`